### PR TITLE
Proposal to add `Isin` operator

### PR DESCRIPTION
**Pull request by cbourjau**
This is a proposal to an `Isin` operator that follows the similarly-named `isin` [function](https://data-apis.org/array-api/latest/API_specification/generated/array_api.isin.html#isin) in the array-api standard. 

**Rendered RFC:** https://github.com/cbourjau/onnx/blob/isin-rfc/docs/proposals/0008-isin.md

### CODE REVIEW

File: docs/proposals/0008-isin.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+- Testing whether an element in `x1` corresponds to an element in `x2` must be determined based on value equality.
+
+### Inputs
+
+**`x1`**: T
+Input tensor whose elements are tested for membership.
+
+**`x2`**: T
+Tensor containing the values to test against.
+
```
[the remark this answers is not in the document]
Comment (cbourjau) [reply docs/proposals/0008-isin.md:56]: It seems like LLMs are very hung up on this. Given the one-sentence semantic of this operator as "Tests for each element in `x1` whether the element is in `x2`." it seems obvious to me that the shape of `x2` does not matter.

**Comment by justinchuby** [comment]
Same question being what models would use it and what scenarios does it unblock? Asking since the array api surface can be fairly large.

**Comment by cbourjau** [comment]
I made the use case in the _Motivation_ section a bit more concrete. We encounter this, for instance, when looking at different coverage options in insurances. 

> Asking since the array api surface can be fairly large.

It is, but this is actually one of the _very_ few remaining functions that ndonnx does not yet fully map to ONNX operators.

**Review by justinchuby — COMMENTED** [review]
Looks useful for filtering. So I support inclusion of this operator.

File: docs/proposals/0008-isin.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+### Type Constraints
+
+**`T`**: tensor(uint8), tensor(uint16), tensor(uint32), tensor(uint64), tensor(int8), tensor(int16), tensor(int32), tensor(int64), tensor(float16), tensor(float), tensor(double), tensor(bfloat16), tensor(string), tensor(bool)
+```
+
+## Drawbacks
+
+This adds another operator to the standard, which needs to be tested, implemented, and maintained by downstream runtimes.
+However, the operator is fairly easy to implement, and there are only a few corner cases to test.
```
Comment (gramalingam) [inline docs/proposals/0008-isin.md:70]: Is there some motivating use-case? Any particular models? Or, is this more useful in an ONNXML setting (classical ML)?

Apart from implementation/maintenance costs, there is also a binary-size cost. It may sound silly, but some users of systems (like onnxruntime) care a lot about binary size (eg., if it is deployed in mobile phone apps). So, there is always a concern about adding something that is not used.

File: docs/proposals/0008-isin.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+- Testing whether an element in `x1` corresponds to an element in `x2` must be determined based on value equality.
+
+### Inputs
+
+**`x1`**: T
+Input tensor whose elements are tested for membership.
+
+**`x2`**: T
+Tensor containing the values to test against.
```
Comment (gramalingam) [inline docs/proposals/0008-isin.md:55]: Is there a shape constraint on x2? It seems reasonable to assume that it is a 1D tensor. (What about x1?)

It's going to be relevant if we define it as a function in onnx. (It's just a matter of avoiding extra Reshapes inside  if it is unnecessary.)

File: docs/proposals/0008-isin.md (hunk 2 above)
Comment (cbourjau) [reply docs/proposals/0008-isin.md:70]: We use it in a "traditional ml" context. The models are proprietary, but they are essentially compositions of sklearn pipelines. It is useful for feature engineering tasks.

Binary size is of course a  runtime issue which depends on the particular circumstances. For most compiled runtimes I assume that everything needed by this operator is already in the binary (iteration and equal comparison). I dont think this will make a dent.

File: docs/proposals/0008-isin.md (hunk 3 above)
Comment (cbourjau) [reply docs/proposals/0008-isin.md:55]: I think constraining either is a net-negative. Being able to have x1 of arbitrary rank seems handy. Forcing users to flatten x2 explicitly is not making the operator implementation easier (iteration over the elements with early stopping). However, runtimes are not guaranteed to have near zero-cost reshaping operations (I recently had to discover this the hard way in onnxruntime).

File: docs/proposals/0008-isin.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+<!--- SPDX-License-Identifier: Apache-2.0 -->
+- Feature Name: Add `Isin` operator to `ai.onnx`
+- Start Date: 2026-08-06
+- RFC PR: [onnx/onnx#8267](https://github.com/onnx/onnx/pull/8267)
+- Status: under discussion
+- Authors:
+  - cbourjau
+
+
+## Add an `Isin` operator to the ai.onnx domain
```
Comment (gramalingam) [inline docs/proposals/0008-isin.md:14]: nit: "IsIn" would be a better name/casing

**Comment by postrational** [comment]
This is a well reasoned proposal and other than binary size I don't see any reason against including it.

I agree with Rama that using `IsIn` fits our UpperCamelCase style names better.

In the actual operator submission we should define this as a function using the inefficient approach as fallback.

File: docs/proposals/0008-isin.md (hunk 4 above)
Comment (gramalingam) [reply docs/proposals/0008-isin.md:14]: I also suggest adding it to the onnxml domain (ai.onnx.ml). It is a signal to backend implementors where it is expected to be used.

**Review by gramalingam — APPROVED** [review]
LGTM, modulo minor comments added above
